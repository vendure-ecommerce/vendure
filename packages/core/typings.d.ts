// This file is for any 3rd party JS libs which don't have a corresponding @types/ package.

declare module 'opn' {
    declare const opn: (path: string) => Promise<any>;
    export default opn;
}

declare module 'i18next-icu' {
    // default
}

declare module 'i18next-fs-backend' {
    // default
}
