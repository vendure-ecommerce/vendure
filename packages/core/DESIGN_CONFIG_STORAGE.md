Design doc created after some back and forth with Claude Code.

All the details are not 100%, but the general direction is there as a good basis:

## Core Components

### Settings Store Field definition

```ts
// Simplified field configuration for MVP
interface SettingsStoreFieldConfig {
    name: string;
    scope?: SettingsStoreScopeFunction; // Defaults to SettingsStoreScopes.global
    public?: boolean;      // Available via Shop API
    readonly?: boolean;    // Cannot be modified via GraphQL API
    requiresPermission?: Array<Permission | string> | Permission | string;
    validate?: (
        value: any,
        injector: Injector,
        ctx: RequestContext,
    ) => string | LocalizedString[] | void | Promise<string | LocalizedString[] | void>;
}

// Registration interface for plugins
interface SettingsStoreRegistration {
    namespace: string; // e.g., 'dashboard', 'payment', 'shipping'
    fields: SettingsStoreFieldConfig[];
}

// Scope function type - returns string to append to key for scoping
// E.g. to store settings for the active user
type SettingsStoreScopeFunction = (params: {
    key: string;
    value?: any;
    ctx: RequestContext;
}) => string;


// Pre-made scope functions
export const SettingsStoreScopes = {
    global: (): string => '',
    user: ({ ctx }: { ctx: RequestContext }): string => ctx.activeUserId || '',
    channel: ({ ctx }: { ctx: RequestContext }): string => ctx.channelId || '',
    userAndChannel: ({ ctx }: { ctx: RequestContext }): string =>
        `${ctx.activeUserId || ''}:${ctx.channelId || ''}`
};
```

### Storage

Single TypeORM entity handles all storage needs:

**Benefits:**

- âœ… Single table design - simple and efficient
- âœ… Built-in TypeORM integration with existing Vendure patterns
- âœ… Automatic unique constraint handling for scoped keys
- âœ… JSON serialization handled transparently
- âœ… Standard Vendure entity lifecycle (timestamps, etc.)
-

```typescript
// TypeORM Entity for storage
@Entity()
export class SettingsStoreEntry extends VendureEntity {

    @Index()
    @Column()
    key: string;

    @Column('json', { nullable: true })
    value: JSONCompatible<any> | null; // JSON-serialized value

    @Index()
    @Column({ nullable: true })
    scopeKey: string | null; // Result of scope function

    // Unique constraint for key + scopeKey combination
@Index('key_scope_unique', ['key', 'scopeKey'], { unique: true })
}
```

## ðŸ”§ **API Design**

### Service Interface

Simplified service interface for MVP:

```typescript

@Injectable()
export class SettingsStoreService {
    /**
     * Register key-value fields (typically called during VendureConfig processing)
     */
    register(registration: SettingsStoreRegistration): void;

    /**
     * Get value with automatic scoping and permission check
     * Returns undefined if key doesn't exist or user lacks permission
     */
    async get<T = any>(key: string, ctx: RequestContext): Promise<T | undefined>;

    /**
     * Set value with automatic scoping and permission check
     * Value must be JSON-serializable
     */
    async set<T = any>(key: string, value: T, ctx: RequestContext): Promise<void>;

    /**
     * Get multiple values efficiently
     */
    async getMany(keys: string[], ctx: RequestContext): Promise<Record<string, any>>;

    /**
     * Set multiple values in a transaction
     */
    async setMany(values: Record<string, any>, ctx: RequestContext): Promise<void>;

    /**
     * Get field definition for a key
     */
    getFieldDefinition(key: string): SettingsStoreFieldConfig | undefined;

    /**
     * Validate a value against its field definition
     */
    async validateValue(
        key: string,
        value: any,
        ctx: RequestContext
    ): Promise<string | LocalizedString[] | void>;
}
```

### Plugin Integration

Plugin registration via VendureConfig:

```typescript
// Dashboard plugin registration for user-specific UI settings
@VendurePlugin({
    // ... other plugin config
    configuration: (config: VendureConfig) => {
        // Register key-value fields via main VendureConfig
        config.settingsStoreFields = {
            ...config.settingsStoreFields,
            dashboard: [
                {
                    name: 'theme',
                    scope: SettingsStoreScopes.user,
                    public: true,
                    requiresPermission: [Permission.Authenticated],
                    validate: (value) => {
                        if (!['light', 'dark', 'auto'].includes(value)) {
                            return 'Theme must be light, dark, or auto';
                        }
                    }
                },
                {
                    name: 'tableFilters',
                    scope: SettingsStoreScopes.userAndChannel, // User-specific per channel
                    public: true,
                    requiresPermission: [Permission.Authenticated]
                },
                {
                    name: 'visibleColumns',
                    scope: SettingsStoreScopes.user,
                    public: true,
                    requiresPermission: [Permission.Authenticated]
                },
                {
                    name: 'notifications',
                    scope: SettingsStoreScopes.user,
                    public: true,
                    requiresPermission: [Permission.Authenticated]
                }
            ]
        };
        return config;
    }
})
export class DashboardPlugin {
    // Plugin implementation...
}
```

### GraphQL Schema

Simplified generic GraphQL schema:

```graphql
# Simplified field definition for MVP
type SettingsStoreFieldDefinition {
    name: String!
    public: Boolean!
    readonly: Boolean!
    requiresPermission: [Permission!]
}

# Generic input/output types using JSON
input SettingsStoreInput {
    key: String!
    value: JSON!
}

# Generic queries and mutations
extend type Query {
    """
    Get value for a specific key (automatically scoped)
    """
    getSettingsStoreValue(key: String!): JSON

    """
    Get multiple settings store pairs (each automatically scoped)
    """
    getSettingsStoreValues(keys: [String!]!): JSON
}

extend type Mutation {
    """
    Set a single settings store pair (automatically scoped)
    """
    setSettingsStoreValue(input: SettingsStoreInput!): Boolean!

    """
    Set multiple settings store pairs in a transaction (each automatically scoped)
    """
    setSettingsStoreValues(inputs: [SettingsStoreInput!]!): Boolean!
}
```

### Plugin Usage with Generic GraphQL API

Any plugin can use these generic operations:

```typescript
// Save user preference - automatically scoped to user
const SAVE_THEME = gql`
    mutation SaveTheme($key: String!, $value: JSON!) {
        setSettingsStoreValue(input: { key: $key, value: $value })
    }
`;

await adminClient.query(SAVE_THEME, {
    variables: { key: 'dashboard.theme', value: 'dark' }
});

// Get multiple settings efficiently
const GET_SETTINGS = gql`
    query GetSettings($keys: [String!]!) {
        getSettingsStoreValues(keys: $keys)
    }
`;

const result = await adminClient.query(GET_SETTINGS, {
    variables: { keys: ['dashboard.theme', 'dashboard.tableFilters'] }
});

// Result: { dashboard.theme: 'dark', dashboard.tableFilters: {...} }
```

## ðŸ”’ **Security & Permissions**

### Permission Model

Simplified permission checking for MVP:

```typescript
// Permission checking based on field configuration (similar to custom fields)
export function userHasSettingsStorePermission(
    ctx: RequestContext,
    fieldDef: SettingsStoreFieldConfig
): boolean {
    // Shop API: only public fields
    if (ctx.apiType === 'shop' && fieldDef.public === true) {
        return true;
    }

    // Admin API: check required permissions
    const requiredPermissions = fieldDef.requiresPermission ?? [];
    if (requiredPermissions.length > 0) {
        return ctx.userHasPermissions(requiredPermissions);
    }

    // For fields without explicit permissions, require basic authentication
    // More granular permission logic can be implemented based on scope function behavior
    return ctx.isAuthorized();
}
```

## ðŸš€ **Implementation Plan**

### MVP (v3.4.0) - Core Settings Store

- [x] Create `SettingsStoreEntry` TypeORM entity with `scope` column
- [x] Implement `SettingsStoreService` with basic CRUD operations
- [x] Add flexible scoping logic using scope functions
- [x] Implement pre-made scope functions (`global`, `user`, `channel`, `userAndChannel`)
- [x] Add permission checking with `requiresPermission` support
- [x] Create GraphQL schema and resolvers
- [x] Add `settingsStoreFields` property to VendureConfig
- [x] Process `settingsStoreFields` during config initialization
- [x] JSON serialization/deserialization
- [x] Basic validation support

### Future Enhancements

- [ ] Type safe inputs and return types, like with custom fields
- [ ] Admin UI components
- [ ] Advanced validation and type safety

## ðŸŽ¨ **Usage Examples**

### User-Specific Dashboard Settings

```typescript
// 1. Register dashboard fields via VendureConfig
const config: VendureConfig = {
    // ... other config
    settingsStoreFields: {
        dashboard: [
            {
                name: 'theme',
                scope: SettingsStoreScopes.user,
                public: true,
                requiresPermission: [Permission.Authenticated],
                validate: (value) => {
                    if (!['light', 'dark', 'auto'].includes(value)) {
                        return 'Theme must be light, dark, or auto';
                    }
                }
            }
        ]
    }
};

// 2. Save user preference - automatically scoped to current user
await settingsStoreService.set('dashboard.theme', 'dark', ctx);

// 3. Retrieve on dashboard load - automatically scoped to current user  
const userTheme = await settingsStoreService.get('dashboard.theme', ctx);

// 4. Update via generic GraphQL mutation
mutation
{
    setSettingsStoreValue(input
:
    {
        key: "dashboard.theme"
        value: "dark"
    }
)
}
```

### Complex Data Storage

```typescript
// Save complex JSON data - all JSON-serializable
const tableFilters = {
    productList: {
        categories: ['electronics'],
        priceRange: { min: 100, max: 500 },
        inStock: true
    },
    orderList: {
        status: ['PaymentSettled', 'PartiallyFulfilled'],
        dateRange: { start: '2024-01-01', end: '2024-12-31' }
    }
};

await settingsStoreService.set('dashboard.tableFilters', tableFilters, ctx);

// Retrieve complex data
const savedFilters = await settingsStoreService.get<typeof tableFilters>('dashboard.tableFilters', ctx);
```

### Mixed Scoping Example

```typescript
// Register fields with different scopes via VendureConfig
const config: VendureConfig = {
    settingsStoreFields: {
        app: [
            {
                name: 'userTheme',
                scope: SettingsStoreScopes.user,
                public: true,
                requiresPermission: [Permission.Authenticated]
            },
            {
                name: 'channelCurrency',
                scope: SettingsStoreScopes.channel,
                public: false,
                requiresPermission: [Permission.ReadSettings]
            },
            {
                name: 'maintenanceMode',
                scope: SettingsStoreScopes.global,
                public: false,
                requiresPermission: [Permission.SuperAdmin]
            }
        ]
    }
};

// Each key is automatically scoped when accessed
const userTheme = await settingsStoreService.get('app.userTheme', ctx);         // User-scoped
const currency = await settingsStoreService.get('app.channelCurrency', ctx);    // Channel-scoped  
const maintenance = await settingsStoreService.get('app.maintenanceMode', ctx); // Global
```

### Advanced Scoping Examples

```typescript
// User AND channel specific settings (same user, different channels)
const config: VendureConfig = {
    settingsStoreFields: {
        dashboard: [
            {
                name: 'tableFilters',
                scope: SettingsStoreScopes.userAndChannel, // User-specific per channel
                public: true,
                requiresPermission: [Permission.Authenticated]
            }
        ]
    }
};

// Custom scope function example
const customDateBasedScope: SettingsStoreScopeFunction = ({ ctx }) => {
    const now = new Date();
    const quarter = Math.floor(now.getMonth() / 3) + 1;
    return `${ctx.activeUserId || ''}:${now.getFullYear()}Q${quarter}`;
};

// Register field with custom scoping
const config: VendureConfig = {
    settingsStoreFields: {
        analytics: [
            {
                name: 'quarterlyTargets',
                scope: customDateBasedScope, // Custom quarterly user-specific scope
                public: false,
                requiresPermission: [Permission.UpdateSettings]
            }
        ]
    }
};
```
