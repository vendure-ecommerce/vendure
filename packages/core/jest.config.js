module.exports = {
    coverageDirectory: 'coverage',
    moduleFileExtensions: ['js', 'json', 'ts'],
    preset: 'ts-jest',
    rootDir: __dirname,
    roots: ['<rootDir>/src', '<rootDir>/mock-data'],
    transform: {
        '^.+\\.(t|j)s$': 'ts-jest',
    },
    globals: {
        'ts-jest': {
            tsconfig: {
                allowJs: true,
            },
        },
    },
    testEnvironment: 'node',
};
