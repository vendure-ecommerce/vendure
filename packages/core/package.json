{
  "name": "@vendure/core",
  "version": "0.5.1",
  "description": "A modern, headless ecommerce framework",
  "repository": {
    "type": "git",
    "url": "https://github.com/vendure-ecommerce/vendure/"
  },
  "keywords": [
    "vendure",
    "ecommerce",
    "headless",
    "graphql",
    "typescript"
  ],
  "private": false,
  "license": "MIT",
  "scripts": {
    "tsc:watch": "tsc -p ./build/tsconfig.build.json --watch",
    "gulp:watch": "gulp -f ./build/gulpfile.ts watch",
    "build": "rimraf dist && tsc -p ./build/tsconfig.build.json && tsc -p ./build/tsconfig.cli.json && gulp -f ./build/gulpfile.ts build",
    "watch": "concurrently yarn:tsc:watch yarn:gulp:watch",
    "lint": "tslint --fix --project ./",
    "test": "jest --config ./jest.config.js",
    "e2e": "jest --config ./e2e/config/jest-e2e.json --runInBand"
  },
  "publishConfig": {
    "access": "public"
  },
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "bin": {
    "vendure": "dist/cli/vendure-cli.js"
  },
  "files": [
    "dist/**/*"
  ],
  "dependencies": {
    "@nestjs/common": "6.8.5",
    "@nestjs/core": "6.8.5",
    "@nestjs/graphql": "6.5.3",
    "@nestjs/microservices": "^6.8.5",
    "@nestjs/platform-express": "6.8.5",
    "@nestjs/testing": "6.8.5",
    "@nestjs/typeorm": "6.2.0",
    "@types/fs-extra": "^8.0.1",
    "@vendure/common": "^0.5.1",
    "apollo-server-express": "2.9.7",
    "bcrypt": "^3.0.6",
    "body-parser": "^1.18.3",
    "chalk": "^2.4.2",
    "commander": "^3.0.2",
    "cookie-session": "^2.0.0-beta.3",
    "csv-parse": "^4.6.5",
    "express": "^4.16.4",
    "fs-extra": "^8.0.1",
    "graphql": "^14.5.8",
    "graphql-iso-date": "^3.6.1",
    "graphql-tag": "^2.10.0",
    "graphql-tools": "^4.0.6",
    "graphql-type-json": "^0.3.0",
    "http-proxy-middleware": "^0.20.0",
    "i18next": "^18.0.1",
    "i18next-express-middleware": "^1.7.1",
    "i18next-icu": "^1.0.1",
    "i18next-node-fs-backend": "^2.1.1",
    "image-size": "^0.8.3",
    "mime-types": "^2.1.21",
    "ms": "^2.1.1",
    "nanoid": "^2.1.6",
    "progress": "^2.0.3",
    "reflect-metadata": "^0.1.12",
    "rxjs": "^6.3.3",
    "typeorm": "0.2.20"
  },
  "devDependencies": {
    "@types/bcrypt": "^3.0.0",
    "@types/cookie-session": "^2.0.36",
    "@types/csv-parse": "^1.1.11",
    "@types/express": "^4.0.39",
    "@types/faker": "^4.1.7",
    "@types/graphql-iso-date": "^3.3.1",
    "@types/graphql-type-json": "^0.3.2",
    "@types/gulp": "^4.0.5",
    "@types/http-proxy-middleware": "^0.19.2",
    "@types/jest": "^24.0.21",
    "@types/mime-types": "^2.1.0",
    "@types/ms": "^0.7.30",
    "@types/nanoid": "^2.1.0",
    "@types/node": "^10.12.18",
    "@types/node-fetch": "^2.5.2",
    "@types/progress": "^2.0.3",
    "@types/prompts": "^2.0.2",
    "gulp": "^4.0.0",
    "mysql": "^2.16.0",
    "node-fetch": "^2.6.0",
    "node-libcurl": "^2.0.2",
    "pg": "^7.8.0",
    "rimraf": "^3.0.0",
    "sql.js": "^1.0.0",
    "sqlite3": "^4.0.6",
    "typescript": "~3.6.2"
  }
}
