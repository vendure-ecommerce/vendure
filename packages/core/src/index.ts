export { bootstrap, bootstrapWorker } from './bootstrap';
export { generateMigration, revertLastMigration, runMigrations } from './migrate';
export * from './api/index';
export * from './common/index';
export * from './config/index';
export * from './event-bus/index';
export * from './health-check/index';
export * from './job-queue/index';
export * from './plugin/index';
export * from './process-context/index';
export * from './entity/index';
export * from './data-import/index';
export * from './service/index';
export * from './i18n/index';
export * from '@vendure/common/lib/shared-types';
export {
    Permission,
    LanguageCode,
    CurrencyCode,
    AssetType,
    AdjustmentType,
} from '@vendure/common/lib/generated-types';
