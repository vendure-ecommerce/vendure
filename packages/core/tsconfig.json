{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "moduleResolution": "NodeNext",
    "declaration": true,
    "removeComments": true,
    "strictPropertyInitialization": false,
    "sourceMap": true,
    "allowJs": true
  }
}
