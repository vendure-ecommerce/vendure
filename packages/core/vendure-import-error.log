Invalid Record Length: header length is 19, got 1 on line 8