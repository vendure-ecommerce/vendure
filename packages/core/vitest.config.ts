import swc from 'unplugin-swc';
import { defineConfig } from 'vitest/config';

export default defineConfig({
    test: {
        /**
         * For local debugging of the unit tests, we set a very long timeout value otherwise tests will
         * automatically fail for going over the 5 second default timeout.
         */
        testTimeout: process.env.TEST_DEBUG ? 1800 * 1000 : 5 * 1000,
        // threads: false,
        // singleThread: true,
        // reporters: ['verbose'],
    },
    plugins: [
        // SWC required to support decorators used in test plugins
        // See https://github.com/vitest-dev/vitest/issues/708#issuecomment-1118628479
        // Vite plugin
        swc.vite({
            jsc: {
                transform: {
                    // See https://github.com/vendure-ecommerce/vendure/issues/2099
                    useDefineForClassFields: false,
                },
            },
        }),
    ],
});
