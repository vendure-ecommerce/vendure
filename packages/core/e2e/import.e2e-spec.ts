import { omit } from '@vendure/common/lib/omit';
import { User } from '@vendure/core';
import { createTestEnvironment } from '@vendure/testing';
import gql from 'graphql-tag';
import path from 'path';

import { initialData } from '../../../e2e-common/e2e-initial-data';
import { testConfig, TEST_SETUP_TIMEOUT_MS } from '../../../e2e-common/test-config';

describe('Import resolver', () => {
    const { server, adminClient } = createTestEnvironment({
        ...testConfig,
        customFields: {
            Product: [
                { type: 'string', name: 'pageType' },
                {
                    name: 'owner',
                    public: true,
                    nullable: true,
                    type: 'relation',
                    entity: User,
                    eager: true,
                },
            ],
            ProductVariant: [{ type: 'int', name: 'weight' }],
        },
    });

    beforeAll(async () => {
        await server.init({
            initialData,
            productsCsvPath: path.join(__dirname, 'fixtures/e2e-products-empty.csv'),
            customerCount: 0,
        });
        await adminClient.asSuperAdmin();
    }, TEST_SETUP_TIMEOUT_MS);

    afterAll(async () => {
        await server.destroy();
    });

    it('imports products', async () => {
        // TODO: waste a few more hours actually fixing this for real
        // Forgive me this abomination of a work-around.
        // On the inital run (as in CI), when the sqlite db has just been populated,
        // this test will fail due to an "out of memory" exception originating from
        // SqljsQueryRunner.ts:79:22, which is part of the findOne() operation on the
        // Session repository called from the AuthService.validateSession() method.
        // After several hours of fruitless hunting, I did what any desperate JavaScript
        // developer would do, and threw in a setTimeout. Which of course "works"...
        const timeout = process.env.CI ? 2000 : 1000;
        await new Promise(resolve => {
            setTimeout(resolve, timeout);
        });

        const csvFile = path.join(__dirname, 'fixtures', 'product-import.csv');
        const result = await adminClient.fileUploadMutation({
            mutation: gql`
                mutation ImportProducts($csvFile: Upload!) {
                    importProducts(csvFile: $csvFile) {
                        imported
                        processed
                        errors
                    }
                }
            `,
            filePaths: [csvFile],
            mapVariables: () => ({ csvFile: null }),
        });

        expect(result.importProducts.errors).toEqual([
            'Invalid Record Length: header length is 17, got 1 on line 8',
        ]);
        expect(result.importProducts.imported).toBe(4);
        expect(result.importProducts.processed).toBe(4);

        const productResult = await adminClient.query(
            gql`
                query GetProducts($options: ProductListOptions) {
                    products(options: $options) {
                        totalItems
                        items {
                            id
                            name
                            slug
                            description
                            featuredAsset {
                                id
                                name
                                preview
                                source
                            }
                            assets {
                                id
                                name
                                preview
                                source
                            }
                            optionGroups {
                                id
                                code
                                name
                            }
                            facetValues {
                                id
                                name
                                facet {
                                    id
                                    name
                                }
                            }
                            customFields {
                                pageType
                                owner {
                                    id
                                }
                            }
                            variants {
                                id
                                name
                                sku
                                price
                                taxCategory {
                                    id
                                    name
                                }
                                options {
                                    id
                                    code
                                }
                                assets {
                                    id
                                    name
                                    preview
                                    source
                                }
                                featuredAsset {
                                    id
                                    name
                                    preview
                                    source
                                }
                                facetValues {
                                    id
                                    code
                                    name
                                    facet {
                                        id
                                        name
                                    }
                                }
                                stockOnHand
                                trackInventory
                                stockMovements {
                                    items {
                                        ... on StockMovement {
                                            id
                                            type
                                            quantity
                                        }
                                    }
                                }
                                customFields {
                                    weight
                                }
                            }
                        }
                    }
                }
            `,
            {
                options: {},
            },
        );

        expect(productResult.products.totalItems).toBe(4);

        const paperStretcher = productResult.products.items.find(
            (p: any) => p.name === 'Perfect Paper Stretcher',
        );
        const easel = productResult.products.items.find((p: any) => p.name === 'Mabef M/02 Studio Easel');
        const pencils = productResult.products.items.find((p: any) => p.name === 'Giotto Mega Pencils');
        const smock = productResult.products.items.find((p: any) => p.name === 'Artists Smock');

        // Omit FacetValues & options due to variations in the ordering between different DB engines
        expect(omit(paperStretcher, ['facetValues', 'options'], true)).toMatchSnapshot();
        expect(omit(easel, ['facetValues', 'options'], true)).toMatchSnapshot();
        expect(omit(pencils, ['facetValues', 'options'], true)).toMatchSnapshot();
        expect(omit(smock, ['facetValues', 'options'], true)).toMatchSnapshot();

        const byName = (e: { name: string }) => e.name;
        const byCode = (e: { code: string }) => e.code;

        expect(paperStretcher.facetValues).toEqual([]);
        expect(easel.facetValues).toEqual([]);
        expect(pencils.facetValues).toEqual([]);
        expect(smock.facetValues.map(byName).sort()).toEqual(['Denim', 'clothes']);

        expect(paperStretcher.variants[0].facetValues.map(byName).sort()).toEqual(['Accessory', 'KB']);
        expect(paperStretcher.variants[1].facetValues.map(byName).sort()).toEqual(['Accessory', 'KB']);
        expect(paperStretcher.variants[2].facetValues.map(byName).sort()).toEqual(['Accessory', 'KB']);
        expect(paperStretcher.variants[0].options.map(byCode).sort()).toEqual(['half-imperial']);
        expect(paperStretcher.variants[1].options.map(byCode).sort()).toEqual(['quarter-imperial']);
        expect(paperStretcher.variants[2].options.map(byCode).sort()).toEqual(['full-imperial']);
        expect(easel.variants[0].facetValues.map(byName).sort()).toEqual(['Easel', 'Mabef']);
        expect(pencils.variants[0].facetValues.map(byName).sort()).toEqual(['Xmas Sale']);
        expect(pencils.variants[1].facetValues.map(byName).sort()).toEqual(['Xmas Sale']);
        expect(pencils.variants[0].options.map(byCode).sort()).toEqual(['box-of-8']);
        expect(pencils.variants[1].options.map(byCode).sort()).toEqual(['box-of-12']);
        expect(smock.variants[0].facetValues.map(byName).sort()).toEqual([]);
        expect(smock.variants[1].facetValues.map(byName).sort()).toEqual([]);
        expect(smock.variants[2].facetValues.map(byName).sort()).toEqual([]);
        expect(smock.variants[3].facetValues.map(byName).sort()).toEqual([]);
        expect(smock.variants[0].options.map(byCode).sort()).toEqual(['beige', 'small']);
        expect(smock.variants[1].options.map(byCode).sort()).toEqual(['beige', 'large']);
        expect(smock.variants[2].options.map(byCode).sort()).toEqual(['navy', 'small']);
        expect(smock.variants[3].options.map(byCode).sort()).toEqual(['large', 'navy']);

        expect(paperStretcher.customFields.owner.id).toBe('T_1');
        expect(easel.customFields.owner.id).toBe('T_1');
        expect(pencils.customFields.owner.id).toBe('T_1');
        expect(smock.customFields.owner.id).toBe('T_1');
    }, 20000);
});
