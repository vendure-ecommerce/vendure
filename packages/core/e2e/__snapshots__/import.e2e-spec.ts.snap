// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import resolver imports products 1`] = `
Object {
  "assets": Array [
    Object {
      "id": "T_1",
      "name": "pps1.jpg",
      "preview": "test-url/test-assets/pps1__preview.jpg",
      "source": "test-url/test-assets/pps1.jpg",
    },
    Object {
      "id": "T_2",
      "name": "pps2.jpg",
      "preview": "test-url/test-assets/pps2__preview.jpg",
      "source": "test-url/test-assets/pps2.jpg",
    },
  ],
  "customFields": Object {
    "owner": Object {
      "id": "T_1",
    },
    "pageType": "default",
  },
  "description": "A great device for stretching paper.",
  "featuredAsset": Object {
    "id": "T_1",
    "name": "pps1.jpg",
    "preview": "test-url/test-assets/pps1__preview.jpg",
    "source": "test-url/test-assets/pps1.jpg",
  },
  "id": "T_1",
  "name": "Perfect Paper Stretcher",
  "optionGroups": Array [
    Object {
      "code": "perfect-paper-stretcher-size",
      "id": "T_1",
      "name": "size",
    },
  ],
  "slug": "perfect-paper-stretcher",
  "variants": Array [
    Object {
      "assets": Array [],
      "customFields": Object {
        "weight": 100,
      },
      "featuredAsset": null,
      "id": "T_1",
      "name": "Perfect Paper Stretcher Half Imperial",
      "price": 4530,
      "sku": "PPS12",
      "stockMovements": Object {
        "items": Array [],
      },
      "stockOnHand": 0,
      "taxCategory": Object {
        "id": "T_1",
        "name": "Standard Tax",
      },
      "trackInventory": "FALSE",
    },
    Object {
      "assets": Array [],
      "customFields": Object {
        "weight": 100,
      },
      "featuredAsset": null,
      "id": "T_2",
      "name": "Perfect Paper Stretcher Quarter Imperial",
      "price": 3250,
      "sku": "PPS14",
      "stockMovements": Object {
        "items": Array [],
      },
      "stockOnHand": 0,
      "taxCategory": Object {
        "id": "T_1",
        "name": "Standard Tax",
      },
      "trackInventory": "FALSE",
    },
    Object {
      "assets": Array [],
      "customFields": Object {
        "weight": 100,
      },
      "featuredAsset": null,
      "id": "T_3",
      "name": "Perfect Paper Stretcher Full Imperial",
      "price": 5950,
      "sku": "PPSF",
      "stockMovements": Object {
        "items": Array [
          Object {
            "id": "T_1",
            "quantity": -10,
            "type": "ADJUSTMENT",
          },
        ],
      },
      "stockOnHand": -10,
      "taxCategory": Object {
        "id": "T_1",
        "name": "Standard Tax",
      },
      "trackInventory": "FALSE",
    },
  ],
}
`;

exports[`Import resolver imports products 2`] = `
Object {
  "assets": Array [],
  "customFields": Object {
    "owner": Object {
      "id": "T_1",
    },
    "pageType": "expanded",
  },
  "description": "Mabef description",
  "featuredAsset": null,
  "id": "T_2",
  "name": "Mabef M/02 Studio Easel",
  "optionGroups": Array [],
  "slug": "mabef-m02-studio-easel",
  "variants": Array [
    Object {
      "assets": Array [],
      "customFields": Object {
        "weight": 300,
      },
      "featuredAsset": null,
      "id": "T_4",
      "name": "Mabef M/02 Studio Easel",
      "price": 91070,
      "sku": "M02",
      "stockMovements": Object {
        "items": Array [
          Object {
            "id": "T_2",
            "quantity": 100,
            "type": "ADJUSTMENT",
          },
        ],
      },
      "stockOnHand": 100,
      "taxCategory": Object {
        "id": "T_1",
        "name": "Standard Tax",
      },
      "trackInventory": "FALSE",
    },
  ],
}
`;

exports[`Import resolver imports products 3`] = `
Object {
  "assets": Array [],
  "customFields": Object {
    "owner": Object {
      "id": "T_1",
    },
    "pageType": "default",
  },
  "description": "Really mega pencils",
  "featuredAsset": null,
  "id": "T_3",
  "name": "Giotto Mega Pencils",
  "optionGroups": Array [
    Object {
      "code": "giotto-mega-pencils-box-size",
      "id": "T_2",
      "name": "box size",
    },
  ],
  "slug": "giotto-mega-pencils",
  "variants": Array [
    Object {
      "assets": Array [
        Object {
          "id": "T_3",
          "name": "box-of-8.jpg",
          "preview": "test-url/test-assets/box-of-8__preview.jpg",
          "source": "test-url/test-assets/box-of-8.jpg",
        },
      ],
      "customFields": Object {
        "weight": 200,
      },
      "featuredAsset": Object {
        "id": "T_3",
        "name": "box-of-8.jpg",
        "preview": "test-url/test-assets/box-of-8__preview.jpg",
        "source": "test-url/test-assets/box-of-8.jpg",
      },
      "id": "T_5",
      "name": "Giotto Mega Pencils Box of 8",
      "price": 416,
      "sku": "225400",
      "stockMovements": Object {
        "items": Array [],
      },
      "stockOnHand": 0,
      "taxCategory": Object {
        "id": "T_1",
        "name": "Standard Tax",
      },
      "trackInventory": "FALSE",
    },
    Object {
      "assets": Array [
        Object {
          "id": "T_4",
          "name": "box-of-12.jpg",
          "preview": "test-url/test-assets/box-of-12__preview.jpg",
          "source": "test-url/test-assets/box-of-12.jpg",
        },
      ],
      "customFields": Object {
        "weight": 200,
      },
      "featuredAsset": Object {
        "id": "T_4",
        "name": "box-of-12.jpg",
        "preview": "test-url/test-assets/box-of-12__preview.jpg",
        "source": "test-url/test-assets/box-of-12.jpg",
      },
      "id": "T_6",
      "name": "Giotto Mega Pencils Box of 12",
      "price": 624,
      "sku": "225600",
      "stockMovements": Object {
        "items": Array [],
      },
      "stockOnHand": 0,
      "taxCategory": Object {
        "id": "T_1",
        "name": "Standard Tax",
      },
      "trackInventory": "FALSE",
    },
  ],
}
`;

exports[`Import resolver imports products 4`] = `
Object {
  "assets": Array [],
  "customFields": Object {
    "owner": Object {
      "id": "T_1",
    },
    "pageType": "default",
  },
  "description": "Keeps the paint off the clothes",
  "featuredAsset": null,
  "id": "T_4",
  "name": "Artists Smock",
  "optionGroups": Array [
    Object {
      "code": "artists-smock-size",
      "id": "T_3",
      "name": "size",
    },
    Object {
      "code": "artists-smock-colour",
      "id": "T_4",
      "name": "colour",
    },
  ],
  "slug": "artists-smock",
  "variants": Array [
    Object {
      "assets": Array [],
      "customFields": Object {
        "weight": 500,
      },
      "featuredAsset": null,
      "id": "T_7",
      "name": "Artists Smock small beige",
      "price": 1199,
      "sku": "10112",
      "stockMovements": Object {
        "items": Array [],
      },
      "stockOnHand": 0,
      "taxCategory": Object {
        "id": "T_2",
        "name": "Reduced Tax",
      },
      "trackInventory": "FALSE",
    },
    Object {
      "assets": Array [],
      "customFields": Object {
        "weight": 500,
      },
      "featuredAsset": null,
      "id": "T_8",
      "name": "Artists Smock large beige",
      "price": 1199,
      "sku": "10113",
      "stockMovements": Object {
        "items": Array [],
      },
      "stockOnHand": 0,
      "taxCategory": Object {
        "id": "T_2",
        "name": "Reduced Tax",
      },
      "trackInventory": "FALSE",
    },
    Object {
      "assets": Array [],
      "customFields": Object {
        "weight": 500,
      },
      "featuredAsset": null,
      "id": "T_9",
      "name": "Artists Smock small navy",
      "price": 1199,
      "sku": "10114",
      "stockMovements": Object {
        "items": Array [],
      },
      "stockOnHand": 0,
      "taxCategory": Object {
        "id": "T_2",
        "name": "Reduced Tax",
      },
      "trackInventory": "FALSE",
    },
    Object {
      "assets": Array [],
      "customFields": Object {
        "weight": 500,
      },
      "featuredAsset": null,
      "id": "T_10",
      "name": "Artists Smock large navy",
      "price": 1199,
      "sku": "10115",
      "stockMovements": Object {
        "items": Array [],
      },
      "stockOnHand": 0,
      "taxCategory": Object {
        "id": "T_2",
        "name": "Reduced Tax",
      },
      "trackInventory": "FALSE",
    },
  ],
}
`;
